_posts/2017-11-23-totw-109.md

```diff
@@ -0,0 +1,111 @@
+---
+title: "Tip of the Week #109: Meaningful `const` in Function Declarations"
+layout: tips
+sidenav: side-nav-tips.html
+published: false
+permalink: tips/109
+type: markdown
+order: "109"
+---
+
+Originally posted as totw/109 on 2016-01-14
+
+By Greg Miller (jgm@google.com)
+
+This document will explain when `const` is meaningful in function declarations,
+and when it is meaningless and best omitted. But first, let us briefly explain
+what is meant by the terms *declaration* and *definition*.
+
+Consider the following code.
+
+```c++
+void F(int);                     // 1: declaration of F(int)
+void F(const int);               // 2: re-declaration of F(int)
+void F(int) { /* ... */ }        // 3: definition of F(int)
+void F(const int) { /* ... */ }  // 4: error: re-definition of F(int)
+```
+
+The first two lines are function *declarations*. A function *declaration* tells
+the compiler the function's signature and return type. In the above example, the
+function's signature is `F(int)`. The constness of the function's parameter type
+is ignored, so both declarations are equivalent ([over.load][CppStandard]).
+
+Lines 3 and 4 from the above code are both function *definitions*. A function
+*definition* is also a declaration, but a definition also contains the
+function's body. Therefore, line 3 is a definition for a function with the
+signature `F(int)`. Similarly, line 4 is also a definition for the same
+function, `F(int)`, which will result in an error at link time. Multiple
+declarations are allowed, but only a single definition is permitted.
+
+Even though the definitions on lines 3 and 4 *declare* and *define* the same
+function, there is a difference within their function bodies due to the way they
+are declared. From the definition on line 3, the type of the function-parameter
+variable within the function will be `int` (i.e., non-const). On the other hand,
+the definition on line 4 will produce a function-parameter variable within the
+function whose type is `const int`.
+
+### Meaningful `const` in function declarations
+
+Not all `const` qualifications in function declarations are ignored. To quote
+from [over.load][CppStandard] in the C++ standard (emphasis added):
```

Comment:
"Overloadable declarations" ([over.load]) in the C++ standard.

Reply:
done